pyroscope/session.go

```diff
@@ -266,126 +270,120 @@ func (ps *Session) uploadData(startTime, endTime time.Time) {
-					Profile:     curMutexBuf,
-					PrevProfile: ps.mutexPrevBytes,
-					SampleTypeConfig: map[string]*upstream.SampleType{
-						"contentions": {
-							DisplayName: "mutex_count",
-							Units:       "lock_samples",
-							Cumulative:  true,
-						},
-						"delay": {
-							DisplayName: "mutex_duration",
-							Units:       "lock_nanoseconds",
-							Cumulative:  true,
-						},
-					},
-				}
-				ps.mergeCumulativeProfile(ps.mergers.Mutex, job)
-				ps.upstream.Upload(job)
-			}
-			ps.mutexPrevBytes = curMutexBuf
-		}
+		ps.dumpMutexProfile(startTime, endTime)
 	}
 
 	if ps.isMemEnabled() {
-		currentGCGeneration := numGC()
-		// sometimes GC doesn't run within 10 seconds
-		//   in such cases we force a GC run
-		//   users can disable it with disableGCRuns option
-		if currentGCGeneration == ps.lastGCGeneration && !ps.disableGCRuns {
-			runtime.GC()
-			currentGCGeneration = numGC()
+		ps.dumpHeapProfile(startTime, endTime)
+	}
+}
+
+func (ps *Session) dumpHeapProfile(startTime time.Time, endTime time.Time) {
+	currentGCGeneration := numGC()
+	// sometimes GC doesn't run within 10 seconds
+	//   in such cases we force a GC run
+	//   users can disable it with disableGCRuns option
+	if currentGCGeneration == ps.lastGCGeneration && !ps.disableGCRuns {
+		runtime.GC()
+		currentGCGeneration = numGC()
+	}
+	if currentGCGeneration != ps.lastGCGeneration {
+		ps.memBuf.Reset()
+		err := ps.deltaHeap.Profile(ps.memBuf)
+		if err != nil {
+			ps.logger.Errorf("failed to dump heap profile: %s", err)
+			return
 		}
-		if currentGCGeneration != ps.lastGCGeneration {
-			pprof.WriteHeapProfile(ps.memBuf)
-			curMemBytes := copyBuf(ps.memBuf.Bytes())
-			ps.memBuf.Reset()
-			if ps.memPrevBytes != nil { //todo does this if statement loose first 10s profile?
-				job := &upstream.UploadJob{
-					Name:        ps.appName,
-					StartTime:   startTime,
-					EndTime:     endTime,
-					SpyName:     "gospy",
-					SampleRate:  100,
-					Format:      upstream.FormatPprof,
-					Profile:     curMemBytes,
-					PrevProfile: ps.memPrevBytes,
-				}
-				ps.mergeCumulativeProfile(ps.mergers.Heap, job)
-				ps.upstream.Upload(job)
-			}
-			ps.memPrevBytes = curMemBytes
-			ps.lastGCGeneration = currentGCGeneration
+		curMemBytes := copyBuf(ps.memBuf.Bytes())
+		job := &upstream.UploadJob{
+			Name:       ps.appName,
+			StartTime:  startTime,
+			EndTime:    endTime,
+			SpyName:    "gospy",
+			SampleRate: 100,
+			Format:     upstream.FormatPprof,
+			Profile:    curMemBytes,
+			SampleTypeConfig: map[string]*upstream.SampleType{
+				"alloc_objects": {
+					Units:      "objects",
+					Cumulative: false,
+				},
+				"alloc_space": {
+					Units:      "bytes",
+					Cumulative: false,
+				},
+				"inuse_space": {
+					Units:       "bytes",
+					Aggregation: "average",
+					Cumulative:  false,
+				},
+				"inuse_objects": {
+					Units:       "objects",
+					Aggregation: "average",
+					Cumulative:  false,
+				},
+			},
```

Comment:
A nit: I understand that this allocation is noting, but if it does not change, a global variable should be fine